bitswap/network/httpnet/httpnet.go

```diff
@@ -0,0 +1,378 @@
+	return nil
+}
+
+func (bsnet *httpnet) Stats() network.Stats {
+	return network.Stats{
+		MessagesRecvd: atomic.LoadUint64(&bsnet.stats.MessagesRecvd),
+		MessagesSent:  atomic.LoadUint64(&bsnet.stats.MessagesSent),
+	}
+}
+
+func (nbsnet *httpnet) TagPeer(p peer.ID, tag string, w int) {
+}
+func (bsnet *httpnet) UntagPeer(p peer.ID, tag string) {
+}
+
+func (bsnet *httpnet) Protect(p peer.ID, tag string) {
+}
+func (bsnet *httpnet) Unprotect(p peer.ID, tag string) bool {
+	return false
+}
+
+func (bsnet *httpnet) NewMessageSender(ctx context.Context, p peer.ID, opts *network.MessageSenderOpts) (network.MessageSender, error) {
+	log.Debugf("NewMessageSender: %s", p)
+	pi := bsnet.host.Peerstore().PeerInfo(p)
+	urls := network.ExtractURLsFromPeer(pi)
+	if len(urls) == 0 {
+		return nil, ErrNoHTTPAddresses
+	}
+
+	return &httpMsgSender{
+		// ctx ??
+		peer:      p,
+		urls:      urls,
+		client:    bsnet.client,
+		receivers: bsnet.receivers,
+		closing:   make(chan struct{}, 1),
+		// opts: todo
+	}, nil
+}
+
+type httpMsgSender struct {
+	client    *http.Client
+	peer      peer.ID
+	urls      []*url.URL
+	receivers []network.Receiver
+	opts      network.MessageSenderOpts
+	closing   chan struct{}
+	closeOnce sync.Once
+}
+
+func (sender *httpMsgSender) SendMsg(ctx context.Context, msg bsmsg.BitSwapMessage) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	bsresp := msg.Clone()
+
+	go func() {
+		for range sender.closing {
+			cancel()
+		}
+	}()
+
+	sendErrors := func(err error) {
+		for _, recv := range sender.receivers {
+			recv.ReceiveError(err)
+		}
+	}
+
+	sendURL, _ := url.Parse(sender.urls[0].String())
+	sendURL.RawQuery = "format=raw&dag-scope=block"
+
+	// TODO: assuming we don't have to manage making concurrent
+	// requests here.
+	for _, entry := range msg.Wantlist() {
+		var method string
+		switch {
+		case entry.Cancel:
+			continue // todo: handle cancelling ongoing
+		case entry.WantType == pb.Message_Wantlist_Block:
+			method = "GET"
+		case entry.WantType == pb.Message_Wantlist_Have:
+			method = "HEAD"
+		default:
+			continue
+		}
+
+		sendURL.Path = "/ipfs/" + entry.Cid.String()
+		headers := make(http.Header)
+		headers.Add("Accept", "application/vnd.ipld.raw")
+
+		req, err := http.NewRequestWithContext(ctx,
+			method,
+			sendURL.String(),
+			nil,
+		)
+		if err != nil {
+			log.Error(err)
+			break
+		}
+		req.Header = headers
+		log.Debugf("cid request to %s %s", method, sendURL)
+		resp, err := sender.client.Do(req)
+		if err != nil { // abort talking to this host
+			log.Error(err)
+			// send error?
+			break
+		}
+		if resp.StatusCode == http.StatusNotFound {
+			if entry.SendDontHave {
+				bsresp.AddDontHave(entry.Cid)
+			}
+			continue
+		}
+
+		// TODO: fixme: limited reader
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			log.Error(err)
+			break
+		}
+
+		if resp.StatusCode != http.StatusOK {
+			err := fmt.Errorf("%s -> %d: %s", sendURL, resp.StatusCode, string(body))
```

Comment:
nit: escape random bytes from random people

```suggestion
			err := fmt.Errorf("%s -> %d: %q", sendURL, resp.StatusCode, string(body))
```